File: packages/python/btw_bridge/src/btw_bridge/pipeline.py
from __future__ import annotations

from btw_bridge.envelope import DeviceEnvelope, EnvelopeFactory
from btw_bridge.sinks.base import EnvelopeSink


class BridgePipeline:
    def __init__(self, factory: EnvelopeFactory, sinks: list[EnvelopeSink]) -> None:
        self._factory = factory
        self._sinks = list(sinks)

    async def forward(self, envelope: DeviceEnvelope) -> None:
        for s in self._sinks:
            await s.emit(envelope)

    async def close(self) -> None:
        for s in self._sinks:
            await s.aclose()

File: packages/nodejs/packages/bridge/src/pipeline.ts
import type { DeviceEnvelope } from "./envelope.js";

export interface EnvelopeSink {
  emit(envelope: DeviceEnvelope): Promise<void>;
  close(): Promise<void>;
}

export class BridgePipeline {
  constructor(private readonly sinks: EnvelopeSink[]) {}

  async forward(envelope: DeviceEnvelope): Promise<void> {
    await Promise.all(this.sinks.map((s) => s.emit(envelope)));
  }

  async close(): Promise<void> {
    await Promise.all(this.sinks.map((s) => s.close()));
  }
}
